libcontainer/container_linux.go

```diff
@@ -2863,3 +2883,63 @@ func needUidShiftOnBindSrc(mount *configs.Mount, config *configs.Config) (bool,
 
 	return true, nil
 }
+
+// Checks if the file at the given path is a bind-mount; if so, returns true and
+// the path to the bind-mount's source.
+func fileIsBindMount(mounts []*mount.Info, fpath string) (bool, string, error) {
+	var fpathMi *mount.Info
+
+	// Since path corresponds to a file (not a directory), if it's a mountpoint
+	// then it must be a bind-mount (i.e., file mountpoints are only allowed for
+	// bind mounts).
+	for _, mi := range mounts {
+		if mi.Mountpoint == fpath {
+			fpathMi = mi
+			break
+		}
+	}
+
+	// If file is not a mountpoint, we are done
+	if fpathMi == nil {
+		return false, "", nil
+	}
+
+	// Find the source of that bind mount. This is not as simple as looking at
```

Comment:
Wondering if this heuristic is fully consistent. For example, what if the 'root-mount' is outside the container's rootfs (i.e., is a bind-mount from the host)?

Reply:
It will always be outside the container's rootfs. It's always a host file that is mounted into the container's rootfs, but the host file itself is a bind-mount that comes from some other file. The algorithm determines where that other file is, so it can mount shiftfs on it and then bind-mount it into the container's rootfs.

Reply:
I guess you're looking at the `/proc/pid/mountinfo` of the user-process hosted within the sys-container. My point was: what if the 'parent' mountinfo entry is not in this mount namespace? For example:

1) user does this at host-level: `mount -o bind file1 file2`
2) user bind-mounts file2 into container's file3